Fixed #28586 -- Added model field lazy loading modes.

adamchainz opened the pull request.

adamchainz commented:
@charettes I’ve been slowly but steadily chipping away at this. Would you like to make an initial review, particularly of the public API in the release notes, and the way that the callable interface works?

charettes commented:
@adamchainz sure! Can it wait this weekend? I'm taking a long break for Easter so I should have plenty of focus time.

adamchainz commented:
Of course, take your time.

charettes reviewed (commented):
👋 @adamchainz I just want to start by saying thank you for the continued work here, this is really looking promising 🙇 

I discussed an idea on how defining a form of fetching _protocol_ (like how _resolvable_ w/ `resolve_expression` and _compilable_ w/ `as_sql` work) to avoid unnecessary closures and keep fetcher modes to the strict minimum which is to receive an instance, a fetcher, and route the call to the fetcher.

There also seems to be a missing usage of fetch mode in `ReverseOneToOneDescriptor.__get__` but maybe this is expected for now as this is still a WIP.

Lastly I left some comments with regards to naming. I'm sure others will have opinions as well but it felt like `django.db.models.fetching` and `fetching_mode` vs `lazy` might be a better .

adamchainz commented on django/db/models/fields/related_descriptors.py:
```diff
@@ -509,16 +517,34 @@ def __get__(self, instance, cls=None):
-                    rel_obj = self.get_queryset(instance=instance).get(**filter_args)
-                except self.related.related_model.DoesNotExist:
-                    rel_obj = None
+                peers = [
+                    peer
+                    for weakref_peer in instance._state.peers
+                    if (peer := weakref_peer()) is not None and not self.is_cached(peer)
+                ]
+                if len(peers) > 1:
+                    # If the instance was fetched with other instances then
+                    # prefetch the field for all of them
+                    qs = self.get_queryset()
+                    prefetch_related_objects(
+                        peers, Prefetch(self.related.get_accessor_name(), queryset=qs)
+                    )
+                    rel_obj = self.related.get_cached_value(instance)
```
Yes, I just didn’t update this yet from Andreas’ original PR copying from django-auto-prefetch, yet.

adamchainz commented on django/db/models/lazy.py:
```diff
@@ -0,0 +1,63 @@
+from contextlib import contextmanager
```
I tried describing the feature name a few different ways before settling on “lazy-loading modes” as in the release notes. But “lazy-fetching modes” could also be a candidate.

Happy to rename the module, but I’d prefer to keep it aligned with the name of the feature.

I agree that “lazy” is used in many ways - translations, “lazy relationships” for passing a string to Foreign Keys, `SimpleLazyObject`... But it does also accurately describes how the data is lazily fetched by a field on access. Naming things is hard.

adamchainz replied:
I’ve tried that for now, and updated the release notes to avoid the word lazy. I went for “on-demand fetching”, which makes sense to me. WDYT?

charettes replied:
I think your proposed structure and wording under `fields.fetching` makes a lot of sense. Thank you for doing that!

adamchainz commented on django/db/models/lazy.py:
```diff
@@ -0,0 +1,63 @@
+    pass
+
+
+def RAISE(instance, field, fetch_for_instances, **kwargs):
```
I was thinking of that, but then maybe users would want “fetch peers” + warnings, or log messages directly rather than warnings. Maybe something we can instead document the pattern for? Like:

```python
def warn_then_fetch_peers(...):
    logger.warning("fetching peers!")
    return FETCH_PEERS(...)

set_default_lazy_mode(warn_then_fetch_peers)
```

charettes replied:
I see, that makes sense to separate non-terminal _surfacing_ of _leaks_ from the actual fetching implementation. I wonder if there is anything that we could provide to avoid such boiler plate though.

adamchainz commented on django/db/models/query_utils.py:
```diff
@@ -203,7 +204,22 @@ def __get__(self, instance, cls=None):
                     )
-                instance.refresh_from_db(fields=[field_name])
+
+                def fetch_for_instances(instances):
+                    if len(instances) == 1:
+                        instances[0].refresh_from_db(fields=[field_name])
+                    else:
+                        attname = self.field.attname
+                        value_by_pk = {
+                            pk: value
+                            for pk, value in self.field.model._base_manager.db_manager()
+                            .filter(pk__in={i.pk for i in instances})
+                            .values_list("pk", attname)
+                        }
+                        for instance in instances:
+                            setattr(instance, attname, value_by_pk[instance.pk])
+
+                get_lazy_mode()(instance, self.field, fetch_for_instances)
```
Yeah, that would be good. I originally tried an interface but it was hard for some reason. At least, it was confusing whether methods should live on the descriptors or fields. I will give it another try, making the descriptors the “fetchers”.

Fetch one and many have separate implementations here only for backwards compatibility. Since its introduction, `refresh_from_db()` [has been documented](https://docs.djangoproject.com/en/5.0/ref/models/instances/#django.db.models.Model.refresh_from_db) as the backing method for deferred attributes:

> Note that when deferred fields are accessed, the loading of the deferred field’s value happens through this method.

There are more backwards compatibility concerns for related descriptors, but there won’t be for `GenericForeignKey` (TODO), so maybe we should only have one method.

charettes replied:
That looks great! Any reason for going with a single `fetch` method over two distinct `fetch_(one|many)`?

I don't see it as a hard blocker but the fact that all `fetch` implementations have an `if` branch that is unnecessary from the caller's perspective (the mode handler knows which method it should call) seemed odd to me.

adamchainz replied:
Now using `fetch_one` and `fetch_many`! No need for a branch.

adamchainz reviewed (commented):
Thank you Simon, thoughtful as always. (P.S. just finished your Django Chat episode yesterday, nice one!). I’ve replied inline to your comments and will work on an update.

> There also seems to be a missing usage of fetch mode in `ReverseOneToOneDescriptor.__get__` but maybe this is expected for now as this is still a WIP.

Yes, this needs doing, as does `GenericForeignKey`.

adamchainz commented:
> Yes, this needs doing, as does `GenericForeignKey`.

I tried `GenericForeignKey` just now, but it’s complicated:

https://github.com/django/django/blob/ec8552417df51df8482df61b8ad78a7002634011/django/contrib/contenttypes/fields.py#L234-L267

A `fetch()` function would need to perform all this validation per-instance. I think for ease of coding and review, we should tackle it in a follow-up PR as an optimization.

charettes commented:
This is looking great @adamchainz thanks for pushing this through and sorry for my slow replies.

If you'd like I could give a shot at implementing the `GenericForeignKey` part in a separate branch that targets this one. Even if it doesn't land in the same PR it would be great to validate that our approach works properly and possibly dog-food the newly introduce [GenericPrefetch](https://docs.djangoproject.com/en/5.0/ref/contrib/contenttypes/#django.contrib.contenttypes.prefetch.GenericPrefetch).

charettes commented:
Completely untested code but from eye balling the other `fetch` methods the following should work without too much issues?

```
diff --git a/django/contrib/contenttypes/fields.py b/django/contrib/contenttypes/fields.py
index 770f88265c..2b7597623b 100644
--- a/django/contrib/contenttypes/fields.py
+++ b/django/contrib/contenttypes/fields.py
@@ -255,16 +255,30 @@ def __get__(self, instance, cls=None):
                 return rel_obj
             else:
                 rel_obj = None
-        if ct_id is not None:
-            ct = self.get_content_type(id=ct_id, using=instance._state.db)
-            try:
-                rel_obj = ct.get_object_for_this_type(
-                    using=instance._state.db, pk=pk_val
-                )
-            except ObjectDoesNotExist:
-                pass
-        self.set_cached_value(instance, rel_obj)
-        return rel_obj
+
+        get_fetching_mode()(self, instance)
+        return self.get_cached_value(instance)
+
+    def fetch(self, instances):
+        f = self.model._meta.get_field(self.ct_field)
+        if len(instances) == 1:
+            ct_id = getattr(instance, f.attname, None)
+            pk_val = getattr(instance, self.fk_field)
+            instance = instances[0]
+            if ct_id is not None:
+                ct = self.get_content_type(id=ct_id, using=instance._state.db)
+                try:
+                    rel_obj = ct.get_object_for_this_type(
+                        using=instance._state.db, pk=pk_val
+                    )
+                except ObjectDoesNotExist:
+                    pass
+            self.set_cached_value(instance, rel_obj)
+        else:
+            prefetch_related_objects(
+                instances, f.name
+            )
+
```

charettes commented on django/db/models/fields/fetching.py:
```diff
@@ -0,0 +1,64 @@
+from asgiref.local import Local
+
+
+def FETCH_ONE(fetcher, instance, **kwargs):
+    fetcher.fetch((instance,))
+
+
+def FETCH_PEERS(fetcher, instance, **kwargs):
+    if instance._state.peers:
+        instances = [
+            peer
+            for weakref_peer in instance._state.peers
+            if (peer := weakref_peer()) is not None
+        ]
+    else:
+        # Peers aren’t tracked for QuerySets returning a single instance
+        instances = (instance,)
+
+    fetcher.fetch(instances)
+
+
+class LazyFieldAccess(Exception):
+    """Blocked lazy access of a model field."""
+
+    pass
+
+
+def RAISE(fetcher, instance, **kwargs):
```
Should we get rid of the `**kwargs`? Were they meant to be used for something?

adamchainz replied:
I left them to leave space for future extensions, if users wrap the calls. But I guess we don’t have that for the callables in `django.db.models.deletion`, so we can drop them.

tolomea replied:
I'm not that familiar with the back compat policy but the kwargs feels unnecessary
if I had a custom fetch mode I would understand if a feature release broke it due to signature change and it would break loud and fast

adamchainz replied:
I got rid of `**kwargs`. If we document “make your own mode”, we can encourage using them there.

charettes commented:
Gave a shot at it here https://github.com/adamchainz/django/pull/4

charettes commented on django/db/models/fields/fetching.py:
```diff
@@ -0,0 +1,64 @@
+
+
+@contextmanager
+def lazy_mode(mode):
```
Should we rename `fetching_mode`?

adamchainz replied:
Yes, but I went with `fetch_mode` to keep it shorter.

tolomea commented:
Thanks Adam for working to push this thing to it's final form. I wonder when the 10 year anniversary of our first version is.

I've had a skim, I plan to look at the details a bit more.

I quite like the fetching modes mechanism and name.

I don't know if this has been covered somewhere but separate modes for deferred fields and FK(and friends) makes a lot of sense to me and is probably what I will look to do in my own code.
With the current API I guess it's possible by doing isinstance checks inside a custom mode.
But that's not very obvious or discoverable.

My rationale here is when you fetch an object from the DB, unless you do stuff to say otherwise all the primitive fields are fetched and none of the FK's.
Absence of an FK is the default it means no one explicitly asked for it.
Absence of a primitive field is unusual it means someone explicitly said don't load this.
So I would want a config that automatically fetches the FK's (in bulk) and errors on the primitive fields.

tolomea commented on django/db/models/fields/related_descriptors.py:
```diff
@@ -254,13 +255,9 @@ def __get__(self, instance, cls=None):
-                # cache on the related object to the current instance to avoid
-                # an extra SQL query if it's accessed later on.
-                if not remote_field.multiple:
-                    remote_field.set_cached_value(rel_obj, instance)
```
did this move? or go away?

tolomea commented:
Also further to the above the doco encourages overridding refresh_from_db which makes setting FETCH_PEERS for deferred fields in your project a third party library compatibility risk as they might done that, although it does feel unlikely.

I wonder if it could / should detect a custom refresh_from_db in FETCH_PEERS and... warn? fall back to FETCH_ONE?

Likewise with the other backward compatibility things.

FWIW I looked though old notes and it seems like the OG version was sometime around September 2016 (later than I remember) so this might still make it into core before the 10th anniversary.

adamchainz commented:
Hello all! @pelme pushed me to continue working on this 👍  I have just pushed a new version with these changes:

1. Rebased onto `main` with updates to target Django 6.0.
2. Merged in the branch from @charettes for `GenericForeignKey` support, plus a tweak to use a separate `GenericForeignKeyFetcher` class so it works with the `RAISE` method. (We could also split a descriptor class from `GenericForeignKey` to achieve this, but I’m concerned about backwards-compatibility, despite it not being public API.)
3. Renamed things to cover “fetch modes” (not *fetching*) consistently.
4. Added a database topic guide on fetch modes.

> I don't know if this has been covered somewhere but separate modes for deferred fields and FK(and friends) makes a lot of sense to me and is probably what I will look to do in my own code.
> With the current API I guess it's possible by doing isinstance checks inside a custom mode.
> But that's not very obvious or discoverable.

I think I‘d rather not split into two separate fetch modes/features. That would be harder to explain and maintain. I would rather just allow custom modes to do this, and document how to. But that would require covering lots of internal APIs right now, which is why I’ve held back on it.

> FWIW I looked though old notes and it seems like the OG version was sometime around September 2016 (later than I remember) so this might still make it into core before the 10th anniversary.

🤞 

---

Review would be appreciated. I’m going to try testing this branch against some real projects to check for edge cases.

pelme reviewed (commented):
Amazing work! This looks great! Me, @Majsvaffla, @nip3o and @saeldion have made this review together. We are very much looking forward having fetch modes in Django.

charettes commented on django/db/models/fields/fetch_modes.py:
```diff
@@ -0,0 +1,63 @@
+    _default = mode
+
+
+@contextmanager
+def fetch_mode(mode):
+    validate_mode(mode)
+
+    orig = getattr(_local, "mode", None)
+    _local.mode = mode
+    try:
+        yield
+    finally:
+        if orig is None:
+            del _local.mode
+        else:
+            _local.mode = orig
```
I've brought it before but the fact the fetch mode is a global variable as opposed to attributed to `QuerySet` instances is likely going to make ergonomics quite weird when wanting to mix modes for distinct querysets or enable it only in some cases.

I believe the way database routing is implemented for example where both `QuerySet.using` and routers can be used to determine what database should be used is battle proven approach. If you then want to rely on global state decorators to advise your router decision you can but Django has lived quite well without having core support for that in database routers.

The lack of queryset method / hook prevents usage of this feature in declarative scenarios such as admins, CBV, or anything that takes a `QuerySet` as a parameter (form, serializer).

e.g.

```python
class BookView(ListView):
   queryset = Book.objects.fetch_mode(FETCH_PEERS)


class AuthorManager(Manager):
   def get_queryset(self):
       return super().get_queryset().fetch_mode(RAISE)

class Author(Model):
    objects = AuthorManager()
```

charettes commented on django/db/models/query.py:
```diff
@@ -142,6 +145,16 @@ def __iter__(self):
                 else:
                     setattr(obj, field.name, rel_obj)
 
+            # Create peers, but only for >1 result, to save memory.
+            if first_obj is None:
+                first_obj = obj
+            else:
```
If the fetching mode was bound to the current queryset instance this whole logic could be delegated to the fetcher itself. That would allow the `FetchPeer` logic to handle the `list[weakref]` creation only when needed while avoiding its creation when dealing with other fetching modes.

adamchainz replied:
This is done.

I also added peers during `RawModelIterable`, making fetch modes work for instances retrieved through raw queries.

charettes reviewed (commented):
I left a few comments about  the global state fetcher approach, `__set_name__` manual calls, systematic creation of `peers: list[weakref]`.

adamchainz commented on docs/topics/db/fetch-modes.txt:
```diff
@@ -0,0 +1,83 @@
+    batch, because you’re likely looping over them.
+
+    This mode reduces most cases of the “N+1 queries problem” to two queries.
+
```
I added the detail in the above sentence:

```rst
Such “peer” instances are tracked in a weak reference list.
```

I don't think we need to use an example, but I agree it's an important detail, at least to allay concerns about memory leaking.

adamchainz commented on docs/topics/db/fetch-modes.txt:
```diff
@@ -0,0 +1,83 @@
+
+When accessing model fields that were not loaded as part of the original query,
+Django will fetch that field’s data from the database. You can customize the
+behavior of this fetching with a **fetch mode**.
```
I have added a list of where field modes apply. They do not apply to related managers or `ManyToManyField` fields. There may be a way to get them to work with `.all()`, because that can be prefetched, but I'm not sure about anything else.

adamchainz commented:
> I've brought it before but the fact the fetch mode is a global variable as opposed to attributed to `QuerySet` instances is likely going to make ergonomics quite weird when wanting to mix modes for distinct querysets or enable it only in some cases.

I have just pushed a new version that is implemented solely as a `QuerySet` method.

After working through it, I agree with you, @charettes, that this cuts with the grain of Django. While the global state is convenient, it could easily mess up some expectations in third-party apps.

I thought it would be nice to allow swapping the default fetch mode, but from some experience testing this PR, I’m not 100% sure it’s a great idea. On one project that I tested, using the `RAISE` fetch mode broke migrations because some migration `RunSQL` operations in Wagtail relied on the default `FETCH_ONE`-style behaviour. Opting in individual `QuerySet`s, or models through a custom manager is the safer option, if a little more involved.

I will respond to the other review comments later today, and I also intend to work on some further changes:

- [ ] Nesting: copy a model instance’s fetch mode to related objects fetched through the fetch mode
- [ ] Nesting: copy a model instance’s fetch mode through its related managers
- [ ] Update `select_related()` and `prefetch_related()` docs to refer to fetch modes

adamchainz commented on docs/topics/db/optimization.txt:
```diff
@@ -219,6 +219,15 @@ them:
 * in view code or other layers, possibly making use of
```
Done, and restructured the whole section.

adamchainz commented on docs/topics/db/optimization.txt:
```diff
@@ -219,6 +219,15 @@ them:
+mitigate a lack of :meth:`~django.db.models.query.QuerySet.select_related` or
+:meth:`~django.db.models.query.QuerySet.prefetch_related` usage when it’s hard
+to predict which fields will be accessed.
+
```
Done and I tried to emphasize that `FETCH_PEERS` should be considered first.

adamchainz commented on docs/releases/6.0.txt:
```diff
@@ -37,6 +37,56 @@ compatible with Django 6.0.
+        # Raise a FieldFetchBlocked exception for any on-demand queries
+        # attempted during template rendering
+        with fetch_mode(RAISE):
+            return render(request, "dashboard.html", context)
```
Outdated now as the context manager/decorator and global state is gone, in favour of the per-QS method.

adamchainz commented on docs/topics/db/fetch-modes.txt:
```diff
@@ -0,0 +1,83 @@
+    field for one instance, you probably need it for all instances in the same
+    batch, because you’re likely looping over them.
+
+    This mode reduces most cases of the “N+1 queries problem” to two queries.
```
Done.

adamchainz commented on docs/topics/db/fetch-modes.txt:
```diff
@@ -0,0 +1,83 @@
+            # Raise a FieldFetchBlocked exception for any on-demand queries
+            # attempted during template rendering
+            with fetch_mode(RAISE):
+                return render(request, "dashboard.html", context)
```
Now irrelevant with the queryset-based implementation.

adamchainz commented on docs/topics/db/fetch-modes.txt:
```diff
@@ -0,0 +1,83 @@
+    A function decorator and context manager that sets the fetch mode for a
+    section of code. This mode will be used for all field fetching within the
+    decorated function or context, within the current thread or asynchronous
+    task.
```
I think the new queryset-based API makes it clear which objects and fetches the fetch mode applies to.

adamchainz commented on docs/topics/db/fetch-modes.txt:
```diff
@@ -0,0 +1,83 @@
+===============
+
+Django provides three modes:
+
```
I’ve updated the new docs with an example that’s explained for each built-in mode.

adamchainz commented:
Re @tolomea 

> I don't know if this has been covered somewhere but separate modes for deferred fields and FK(and friends) makes a lot of sense to me and is probably what I will look to do in my own code.

I understand the rationale for splitting behaviour between FKs and deferred fields, but I think actually splitting them in the API would be too fine-grained. Fetching deferred fields in bulk is still useful, such as for [model classes that defer some fields by default](https://adamj.eu/tech/2023/12/11/django-always-defer-a-field/).

> With the current API I guess it's possible by doing isinstance checks inside a custom mode. But that's not very obvious or discoverable.

Yes, this would be the way. We could document a public API around combining fetch modes like this, but maybe in a follow-up PR.

> Also further to the above the doco encourages overridding refresh_from_db which makes setting FETCH_PEERS for deferred fields in your project a third party library compatibility risk as they might done that, although it does feel unlikely.

Thankfully, this risk is gone by adopting a `QuerySet`-based API. Instead, third-party libraries will need to opt into using a different fetch mode.

adamchainz commented on django/db/models/fields/__init__.py:
```diff
@@ -954,7 +954,9 @@ def contribute_to_class(self, cls, name, private_only=False):
-            setattr(cls, self.attname, self.descriptor_class(self))
+            descriptor = self.descriptor_class(self)
+            setattr(cls, self.attname, descriptor)
+            descriptor.__set_name__(cls, self.name)
```
The aim here was to have the "fetcher protocol" include a consistent way to see the related field's name available, so the `RAISE` mode can rely on it.

Descriptors did not consistently point at fields. At least `GenericForeignKey` is, oddly, its own descriptor. Adding a recursive `field` attribute there felt wrong, but it seemed to make sense to call [`__set_name__`](https://docs.python.org/3/reference/datamodel.html#object.__set_name__) here because it's the same point that Python would do so if the descriptor were placed in the class definition.

In retrospect, it's probably less invasive and more sensible to make `field` consistently available, and fix GFK to have a separate descriptor class. I'll give that a try now.

adamchainz replied:
Okay, I just pushed a fix. It required a new commit that splits a descriptor class from `GenericForeignKey`, which I also put into a new PR at #19381. Another minor hack it required is a new property in `ReverseOneToOneDescriptor` for `field`, with a docstring explaining the ducktyping.

adamchainz commented on django/db/models/fields/related_descriptors.py:
```diff
@@ -133,6 +134,9 @@ class Child(Model):
     def __init__(self, field_with_rel):
         self.field = field_with_rel
 
+    def __set_name__(self, owner, name):
+        self.name = name
+
```
Per above - trying with field.

https://github.com/django/django/pull/17554#discussion_r2042249701

adamchainz commented:
I've just pushed an update with a new commit that does the "nesting", copying the fetch mode to related objects to apply to a whole tree of objects unless explicitly replaced.

I'm still going to work a bit on the `select_related` and `prefetch_related` docs, but I think we're nearly there on the `QuerySet`-based design.

adamchainz commented:
I just fixed the test failure. (I wrote *behaviour* instead of *behavior*, failing the spellcheck!)

Zerotask commented on docs/ref/models/querysets.txt:
```diff
@@ -1149,12 +1160,14 @@ that is that ``select_related('foo', 'bar')`` is equivalent to
+Returns a ``QuerySet`` that will automatically retrieve the given lookups, each
+in one extra batch query. Prefetching is a way to optimize database access
+when you know you'll be accessing related objects later, so you can avoid
+triggering triggering the on-demand loading behavior of the model instances’
```
```suggestion
triggering the on-demand loading behavior of the model instances’
```

adamchainz replied:
Thank you for spotting!

g-nie commented on django/db/models/query_utils.py:
```diff
@@ -241,6 +242,20 @@ def _check_parent_chain(self, instance):
+        value_by_pk = {
+            pk: value
+            for pk, value in self.field.model._base_manager.db_manager()
+            .filter(pk__in={i.pk for i in instances})
```
I wonder if the set comprehension here could have some performance impact for a large number of `instances`. I'm not sure what else could be done instead though.

adamchainz replied:
It shouldn't take a meaningful amount of time. Yes, it's for N instances, but gathering their in-memory PKs will be way faster than the actual query.

g-nie reviewed (commented):
Thanks @adamchainz this looks great! I left a small thought

adamchainz commented on docs/topics/db/optimization.txt:
```diff
@@ -196,28 +196,45 @@ thousands of records are returned. The penalty will be compounded if the
-* in view code or other layers, possibly making use of
-  :func:`~django.db.models.prefetch_related_objects` where needed.
+Where :attr:`~django.db.models.query.QuerySet.prefetch_related` would be useful
+“after the fact”, use :func:`~django.db.models.prefetch_related_objects` to
```
Yes, that's much better, thanks!

adamchainz commented on docs/topics/db/queries.txt:
```diff
@@ -1665,8 +1665,14 @@ you, but we point it out here for the curious.)
+
+These accessors may be pre-fetched by the ``QuerySet`` methods
+:meth:`~django.db.models.query.QuerySet.select_related` or
+:meth:`~django.db.models.query.QuerySet.prefetch_related`. If not pre-fetched,
```
Done, thanks!

jacobtylerwalls commented on docs/ref/models/querysets.txt:
```diff
@@ -1149,12 +1160,14 @@ that is that ``select_related('foo', 'bar')`` is equivalent to
-This has a similar purpose to ``select_related``, in that both are designed to
-stop the deluge of database queries that is caused by accessing related objects,
-but the strategy is quite different.
+This method has a similar purpose to :meth:`select_related`, in that both are
+designed to eagerly fetch related objects. However, they work in different ways.
```
I would expect chaining `.prefetch_related(None)` to clear a FETCH_PEERS fetch mode, but it doesn't -- it continues to fetch peers. What is your view on what should happen?

tolomea replied:
currently the two API's are manipulating independent state, it'd feel weird to clear it like this instead of using fetch_mode
of course your other suggestion does couple them in another way

jacobtylerwalls replied:
Makes sense. (To my ear the docs advertised this as the successor to `prefetch_related()`, but looking again the docs clarify that there is a separate use case for `prefetch_related()`, e.g. in managers, so I'm convinced these are really two APIs after all.)

adamchainz replied:
Yeah, they are intended as two separate APIs. I tried to clarify that fetch modes apply to “on-demand fetches”, the existing behaviour where accessing an attribute fetches data. I think it’s clear enough, but if you have any suggested docs tweaks, I’m happy to see them.

tolomea commented on docs/topics/db/optimization.txt:
```diff
@@ -196,28 +196,45 @@ thousands of records are returned. The penalty will be compounded if the
+field access more efficient with bulk-fetching. Enable all it for all usage of
+your models :ref:`with a custom manager <fetch-modes-custom-manager>`.
+
+Using this fetch mode is easier than declaring fields to fetch with
+:meth:`~django.db.models.query.QuerySet.select_related` or
+:meth:`~django.db.models.query.QuerySet.prefetch_related`, especially when it's
+hard to predict which fields will be accessed.
```
that select_related() call is almost always a bad idea, it's very prone to over fetching and to doing so in a way that is fairly hostile to the database, I'd be more in favor of removing that behaviour

adamchainz replied:
Ticket: https://code.djangoproject.com/ticket/36593

charettes commented on django/db/models/fields/related_descriptors.py:
```diff
@@ -441,16 +447,20 @@ def RelatedObjectDoesNotExist(self):
             )
-        queryset = querysets[0] if querysets else self.get_queryset()
+        queryset = (
+            querysets[0] if querysets else self.get_queryset(instance=instances[0])
+        )
         queryset._add_hints(instance=instances[0])
```
Aren't these two `instance` hint passing redundant?

adamchainz replied:
Well spotted, thanks, removed.

charettes reviewed (commented):
Thanks for pushing this forward and sorry for the delay in review @adamchainz.

I left a few comments mainly requesting clarifications and more verbose commit message but in essence the state of this PR is looking great to me 🥇 🙇

jacobtylerwalls commented on django/db/models/query.py:
```diff
@@ -291,6 +308,7 @@ def __init__(self, model=None, query=None, using=None, hints=None):
         self._prefetch_done = False
         self._known_related_objects = {}  # {rel_field: {pk: rel_obj}}
         self._iterable_class = ModelIterable
+        self._fetch_mode = FETCH_ONE
```
When debugging a project earlier this summer, it was enormously helpful to be able to just monkey-patch this to `RAISE` and comb through failures.

One _interesting_ failure I wasn't able to solve looked like:

```py
request.user.userprofile
```
```py
django.core.exceptions.FieldFetchBlocked: Fetching of User.userprofile blocked.
```

To fix that, I'd need:
- a custom user model
- a custom model manager for that custom model that overrides `get_queryset()` to chain `.fetch_mode(models.FETCH_PEERS)`

My questions are:
- that's the easiest way, right?
- Is there a future where `django.contrib.auth` would ever use `.fetch_mode(models.FETCH_PEERS)` in `get_user()`? Since it's a contrib app? Or would that violate the mental model of "django defaults to FETCH_ONE"?

One query for a `.userprofile` per request isn't so bad, but the fact that I can't "globally" turn on RAISE mode without my tests immediately failing on this is actually what impacts me more.

adamchainz replied:
> that's the easiest way, right?

I think it would be slightly easier to subclass `ModelBackend` and replace `get_user` with a version that adds the `.fetch_mode()` call.

> Is there a future where `django.contrib.auth` would ever use `.fetch_mode(models.FETCH_PEERS)` in `get_user()`? Since it's a contrib app? Or would that violate the mental model of "django defaults to FETCH_ONE"?

I think there could be such a future. I'd hope we could also change the admin to use `models.FETCH_PEERS` by default.

> One query for a `.userprofile` per request isn't so bad, but the fact that I can't "globally" turn on RAISE mode without my tests immediately failing on this is actually what impacts me more.

Yeah, I found a similar issue in a project I tried the same trick on. That one failed during migrations, because a data migration in a third-party app was relying on the default `FETCH_ONE` behaviour.

I don't think there's a good API we can provide to allow enabling `RAISE` globally without breaking lots of hard-to-fix stuff. That's why I've come around to Simon's suggestion of going with the `QuerySet`-based API. We can always address adding a supported global switch later, but for now having the targeted API seems best.

jacobtylerwalls commented on docs/releases/6.0.txt:
```diff
@@ -90,6 +90,50 @@ Notably, the return type of the :class:`EmailMessage.message()
 :py:class:`email.message.EmailMessage`. This supports the same API as the
 previous ``SafeMIMEText`` and ``SafeMIMEMultipart`` return types, but is not an
 instance of those now-deprecated classes.
+Model field fetch modes
```
```suggestion

Model field fetch modes
```

jacobtylerwalls reviewed (commented):
I'm looking forward to this one @adamchainz. Couple of usage / roadmap questions for you.

adamchainz commented on django/contrib/contenttypes/fields.py:
```diff
@@ -257,7 +267,10 @@ def __get__(self, instance, cls=None):
-        return rel_obj
+
+    def fetch_many(self, instances):
+        missing_instances = [i for i in instances if not self.field.is_cached(i)]
```
Done, and similarly in `ForwardManyToOneDescriptor` and `ReverseOneToOneDescriptor` as well.

adamchainz commented on django/db/models/fetch_modes.py:
```diff
@@ -0,0 +1,52 @@
+    def fetch(self, fetcher, instance):
+        klass = instance.__class__.__qualname__
+        field_name = fetcher.field.name
+        raise FieldFetchBlocked(f"Fetching of {klass}.{field_name} blocked.")
```
Done, and expanded the tests to check for this with extra assertions like:

```python
with self.assertRaisesMessage(FieldFetchBlocked, msg) as cm:
    ...
self.assertIsNone(cm.exception.__cause__)
self.assertTrue(cm.exception.__suppress_context__)
```

adamchainz commented on docs/topics/db/optimization.txt:
```diff
@@ -196,28 +196,45 @@ thousands of records are returned. The penalty will be compounded if the
+
+It may be useful to apply these methods in :doc:`managers and default managers
+</topics/db/managers>`. Be aware when your manager is and is not used;
+sometimes this is tricky so don't make assumptions.
```
I totally agree, but this wording was pre-existing, and this PR is already giant. Can we punt this one?

adamchainz commented on django/db/models/fields/related_descriptors.py:
```diff
@@ -441,16 +447,20 @@ def RelatedObjectDoesNotExist(self):
 
-    def get_queryset(self, **hints):
-        return self.related.related_model._base_manager.db_manager(hints=hints).all()
+    def get_queryset(self, *, instance):
+        return self.related.related_model._base_manager.db_manager(
+            hints={"instance": instance}
```
Good point, I have split the signature change into a new commit.

adamchainz commented on django/db/models/base.py:
```diff
@@ -595,7 +612,7 @@ def __init__(self, *args, **kwargs):
 
     @classmethod
-    def from_db(cls, db, field_names, values):
+    def from_db(cls, db, field_names, values, *, fetch_mode=None):
```
I think we should leave it as a private API for now. Let's allow the tinkerers who read the source to play with it for the first release.

But at least we've documented two attributes of `Model._state` ([link](https://docs.djangoproject.com/en/5.2/ref/models/instances/#django.db.models.Model._state)) so there's precedent to document it more.

adamchainz commented on django/db/models/query_utils.py:
```diff
@@ -281,6 +282,20 @@ def _check_parent_chain(self, instance):
+        value_by_pk = {
+            pk: value
+            for pk, value in self.field.model._base_manager.db_manager()
+            .filter(pk__in={i.pk for i in instances})
```
After a quick search, I think Django already doesn't support unhashable primary keys. The deletion collector performs a similar set comprehension:

https://github.com/django/django/blob/2722cb61ccae84f593e6d2c28814e3c628743994/django/db/models/deletion.py#L490

Similarly, model forms rely on hashing PKs:

https://github.com/django/django/blob/2722cb61ccae84f593e6d2c28814e3c628743994/django/forms/models.py#L720

Also, because values must be unique in the database, that implies a hash function can be written.

So, I think this isn't a use case we need to worry about.

jacobtylerwalls replied:
Appreciate the research, that's reassuring!

adamchainz commented on docs/topics/db/fetch-modes.txt:
```diff
@@ -0,0 +1,140 @@
+* Fields deferred with :meth:`.QuerySet.defer` or :meth:`.QuerySet.only`
+* :ref:`generic-relations`
+
+Fetch modes are copied to related objects when they are fetched, so they can
```
Done, in a new commit. I wasn't sure if it was worth documenting, since `Model.objects.fetch_mode(...).create(...)` seems like an obvious pattern that should work, and the use case seems kinda niche.

adamchainz commented on django/db/models/fetch_modes.py:
```diff
@@ -0,0 +1,52 @@
+        fetcher.fetch_one(instance)
+
+
+FETCH_ONE = FetchOne()
```
I think initial designs used an enum or constants, but @charettes suggested to use callables, copying the `on_delete` modes like `CASCADE`. Not only is this more consistent, it opens the door for adding custom logic, like including logging or conditionally blocking fetches.

jacobtylerwalls replied:
I'm down with the callables, but I just meant wrapping them in an enum like this, mostly just for discoverability:
```py
class FetchModes(enum.Enum):
    FETCH_ONE = FETCH_ONE
    FETCH_PEERS = FETCH_PEERS
    RAISE = RAISE
```

jacobtylerwalls replied:
No problem. Just picking your brain 👍

adamchainz commented on docs/topics/db/fetch-modes.txt:
```diff
@@ -0,0 +1,140 @@
+apply to a whole tree of relationships, not just the top-level model in the
+initial ``QuerySet``.
+
+.. admonition:: Referencing fetch modes
+
+    Fetch modes are defined in ``django.db.models.fetch_modes``, but for
+    convenience they're imported into :mod:`django.db.models`. The standard
+    convention is to use ``from django.db import models`` and refer to the
+    fetch modes as ``models.<mode>``.
```
Yes, totally! Thanks for spotting this, fixed.

adamchainz commented on docs/releases/6.0.txt:
```diff
@@ -37,6 +37,56 @@ compatible with Django 6.0.
+
+The on-demand fetching behavior of model fields is now configurable with
+:doc:`fetch modes </topics/db/fetch-modes>`. Django provides three fetch
+modes:
```
I've tried editing this paragraph in various ways, but to me, it reads a bit clunky to mention "N+1 queries" here and then very shortly after. I hope we can trust users' attention span to reach point 2, plus we can advertise the feature in other ways, such as blog posts and conference talks.

adamchainz commented on django/db/models/base.py:
```diff
@@ -473,6 +482,14 @@ class ModelState:
+    def __getstate__(self):
+        state = self.__dict__.copy()
+        state.pop("fetch_mode", None)
+        state.pop("peers", None)
```
I ended up adding `__reduce__` methods to fetch modes to make them pickle more efficiently, in a separate commit 30c72657ff9d3747eb5d4b59aacce9bac0acdbad.

adamchainz commented:
Alright, I think I handled all review comments!

jacobtylerwalls commented on django/db/models/query_utils.py:
```diff
@@ -281,6 +282,20 @@ def _check_parent_chain(self, instance):
+        value_by_pk = {
+            pk: value
+            for pk, value in self.field.model._base_manager.db_manager()
+            .filter(pk__in={i.pk for i in instances})
```
Any interest in observing database parameter limits, similar to how `in_bulk()` [does it](https://github.com/django/django/blob/0ddbe12ea99a2dc1b757dc2015ba8bb6bfd9d653/django/db/models/query.py#L1195), or factoring out some reusable private helper?

charettes replied:
Kind of a shame that `in_bulk` doesn't support `values` / `values_list` in the first place as that could have been

```python
value_by_pk = (
    self.field.model._base_manager
    .values_list(attname, flat=True).in_bulk({i.pk for i in instances})
)
```

in all cases it seems that the `db_manager()` call is unnecessary?

jacobtylerwalls replied:
> Kind of a shame that in_bulk doesn't support values / values_list

I'll open a new features ticket and link to the [sketched](https://github.com/django/django/pull/18497) implementation, given that we have this additional use in Django itself now.

charettes replied:
Also, not sure if the dict literal is faster but the `dict` function can be used as well

```python
        value_by_pk = dict(
            self.field.model._base_manager
            .filter(pk__in={i.pk for i in instances})
            .values_list("pk", attname)
        )

adamchainz commented on django/db/models/fetch_modes.py:
```diff
@@ -0,0 +1,61 @@
+from django.core.exceptions import FieldFetchBlocked
+
+
+class FetchMode:
```
I think I will leave this as a follow-up, I haven't found the time to really hash out the details, and the 6.0 merge deadline is here… Also, it's probably best to leave it for tinkerers to explore for one version before we commit to a certain pattern.

I'd also like to make another follow-up, to use `FETCH_PEERS` by default in the admin.

adamchainz commented on django/db/models/fetch_modes.py:
```diff
@@ -0,0 +1,61 @@
+    track_peers = True
+
+    def fetch(self, fetcher, instance):
+        instances = [
+            peer
+            for peer_weakref in instance._state.peers
+            if (peer := peer_weakref()) is not None
+        ]
```
I think best to leave this for a follow-up where we document how to make a custom fetch mode.

adamchainz commented:
Rebased targeting Django 6.1!

jacobtylerwalls commented:
Thanks to all involved! Will merge when checks finish.
